Slim/Utils/Prefs.pm

```diff
@@ -269,6 +269,14 @@ sub init {
 		'noupnp'                => 1,
 		'onlyAlbumYears'        => 1,
 	);
+	if ($prefs->get('useUnifiedArtistsList')) {
+		$defaults{'artistAlbumLink'} = 1 if !$prefs->exists('artistAlbumLink');
+		$defaults{'albumartistAlbumLink'} = 1 if !$prefs->exists('albumartistAlbumLink');
+		$defaults{'trackartistAlbumLink'} = $prefs->get('trackartistInArtists') if !$prefs->exists('trackartistAlbumLink');
+		$defaults{'composerAlbumLink'} = $prefs->get('composerInArtists') if !$prefs->exists('composerAlbumLink');
+		$defaults{'conductorAlbumLink'} = $prefs->get('conductorInArtists') if !$prefs->exists('conductorAlbumLink');
+		$defaults{'bandAlbumLink'} = $prefs->get('bandInArtists') if !$prefs->exists('bandAlbumLink');
+	}
```

Comment:
Oh... initialising flags to positive values is a mess... I tripped over this over and over again... So, I'd say just initialise the values here with `1` unconditionally. And then you'll have to make sure they're set to a well defined `0` when disabling them, so not to leave them undefined. I think you should be able to have an `onChange` handler for them making sure the value was never undefined, but replace them with `0`.

Reply:
I didn't know what was going on: every time I restarted the server, it initialized un-checked boxes again. Don't know how many times I checked all my variable names before figuring it out!

Also, that solution is wrong: if it runs with separate artist lists, nothing gets initialized. Trivial to fix, though.

Shouldn't the prefs backend distinguish between a missing pref and one which exists but is undef?

But, we could get around it be enclosing the new xxxAlbumLink values in a hash, so it's a single pref, much like I did for user-defined roles. (Could also apply this to xxxInArtists, which should be straightforward, though it might force some changes on Material.) 

What do you think?

Reply:
Undef is undef - no matter whether it was actively set to undef or not. Therefore it's code disabling a flag which has to make sure it's not undefined, but a defined falsy value. That's one of the problems with untypted languages like Perl...

Reply:
Yesterday I tried explicitly setting the pref to "0" if the checkbox was unchecked. It didn't seem to work. This morning I discovered that I shouldn't be including prefs that are explicitly set by code in the handler in the returned list - this was causing a second prefs write, overwriting my zeros with undef.

So it's all resolved now, and I've been able to go back to the simpler initialisation/migration code.